### PR TITLE
Feat/lyapunov stubs

### PR DESCRIPTION
**Pull request by SomeUserName1**
# Packaging Summary

This document summarizes the pip-installable package setup created for the `elmneuron` repository.

### 1. `pyproject.toml` (Main Package Configuration)

Modern Python package configuration following PEP 518 and PEP 621 standards.

**Key features:**
- Package metadata (name, version, description, authors)
- Dependencies with optional groups (`dev`, `wandb`, `gpu`, `all`)
- Build system configuration (setuptools)
- Tool configurations (black, isort, pytest)
- Entry points for package discovery

**Dependencies:**
- Core: torch, numpy, h5py, matplotlib, seaborn, scikit-learn, tqdm
- Dev: pytest, black, isort, pre-commit
- Optional: wandb (experiment tracking)

### 2. `src/elmneuron/__init__.py` (Package Entry Point)

Main package initialization file with public API exports.

**Exports:**
- `ELM_v1`: Original ELM implementation
- `ELM_v2`: Branch-ELM with 7x parameter reduction
- Utility functions: `MLP`, routing functions, activation functions
- `__version__`: Package version (0.1.0)

### 3. `src/elmneuron/neuronio/__init__.py`

NeuronIO dataset module initialization.

**Exports:**
- `NeuronIO`: Dataset class
- `NeuronioLoss`: Custom loss function
- `NeuronioEvaluator`: Evaluation utilities
- Data loading and visualization functions

### 4. `src/elmneuron/shd/__init__.py`

Spiking Heidelberg Digits dataset module initialization.

**Exports:**
- `SHD`: Standard SHD dataset
- `SHDAdding`: SHD-Adding variant for long-range dependencies
- Dataset utilities and download functions

### 5. `MANIFEST.in` (Package Data Specification)

Specifies which non-Python files to include in the distribution.

**Included:**
- Documentation: README.md, CLAUDE.md, LICENSE, INSTALL.md
- Environment files: elm_env.yml, elm_env_gpu.yml
- Model images: *.png files
- Pre-trained models: models/*.pt, models/*.md

**Excluded:**
- Test files, notebooks, cache directories
- Git files, wandb logs, downloaded datasets

### 6. `src/elmneuron/py.typed` (Type Hint Marker)

PEP 561 marker file indicating the package supports type hints.

### 7. `INSTALL.md` (Installation Guide)

Comprehensive installation instructions covering:
- Installation from source (development and user modes)
- Conda environment setup
- Wheel building and installation
- GPU support configuration
- Troubleshooting common issues
- Verification steps

### 8. Updated `CLAUDE.md`

Added sections on:
- Code quality and formatting (pre-commit hooks)
- Package building and distribution
- Corrected import paths (`elmneuron` not `src`)
- Git workflow conventions
- Additional development notes

## Package Structure

```
elmneuron/
├── pyproject.toml              # Modern package configuration
├── MANIFEST.in                 # Non-Python file inclusion rules
├── INSTALL.md                  # Installation guide
├── README.md                   # Project documentation
├── CLAUDE.md                   # Claude Code guidance (updated)
├── LICENSE                     # MIT license
├── src/
│   └── elmneuron/              # Main package
│       ├── __init__.py         # Package entry point
│       ├── py.typed            # Type hint marker
│       ├── expressive_leaky_memory_neuron.py       # ELM v1
│       ├── expressive_leaky_memory_neuron_v2.py    # ELM v2
│       ├── modeling_utils.py   # Shared utilities
│       ├── neuronio/           # NeuronIO dataset
│       │   └── __init__.py     # Submodule exports
│       └── shd/                # SHD dataset
│           └── __init__.py     # Submodule exports
├── tests/                      # Test suite
├── notebooks/                  # Jupyter notebooks (not packaged)
├── models/                     # Pre-trained models (included in wheel)
└── dist/                       # Built distributions
    └── elmneuron-0.1.0-py3-none-any.whl
```

### For Development

```bash
# Clone repository
git clone git@github.com:AaronSpieler/elmneuron.git
cd elmneuron

# Install in editable mode with dev tools
pip install -e ".[dev]"

# Set up pre-commit hooks
pre-commit install
```

### For Users

```bash
# From source
pip install .

# Or with optional dependencies
pip install ".[wandb]"

# From wheel (after building)
pip install dist/elmneuron-0.1.0-py3-none-any.whl
```

### Using Conda

```bash
# Create environment
conda env create -f elm_env.yml
conda activate elm_env

# Install package
pip install -e .
```

## Building the Package

```bash
# Install build tool
pip install build

# Build wheel (recommended)
python -m build --wheel

# Build both wheel and source distribution
python -m build

# Output in dist/ directory
```

## Verification

Test the installation:

```python
import elmneuron
from elmneuron import ELM_v1, ELM_v2

print(f"Version: {elmneuron.__version__}")  # 0.1.0

# Create models
model_v2 = ELM_v2(num_input=128, num_output=2, num_memory=15)
model_v1 = ELM_v1(num_input=128, num_output=2, num_memory=20)

# Use dataset loaders
from elmneuron.neuronio import NeuronIO, NeuronioLoss
from elmneuron.shd import SHD, SHDAdding
```

## Pre-commit Hooks

The package includes pre-commit hook configuration:

```bash
# Install hooks
pre-commit install

# Run manually
pre-commit run --all-files
```

**Hooks configured:**
- `black`: Code formatting (88 char line length)
- `isort`: Import sorting (black-compatible)
- `check-ast`: Python syntax validation
- `check-yaml`: YAML file validation
- `trailing-whitespace`: Remove trailing whitespace
- `end-of-file-fixer`: Ensure newline at EOF

## Version Management

To release a new version:

1. Update version in `pyproject.toml` (line 7)
2. Update `__version__` in `src/elmneuron/__init__.py` (line 25)
3. Build new wheel: `python -m build --wheel`
4. Tag release in git: `git tag v0.x.x`

## Publishing to PyPI

When ready to publish (requires PyPI account and twine):

```bash
# Build distributions
python -m build

# Upload to TestPyPI first (recommended)
python -m twine upload --repository testpypi dist/*

# Upload to PyPI
python -m twine upload dist/*
```

## Testing the Package

```bash
# Run tests
pytest tests/

# Run specific test
pytest tests/test_elm_models.py::TestInitialization -v

# Run with coverage
pytest --cov=elmneuron tests/
```

## Key Design Decisions

1. **Src-layout**: Uses `src/elmneuron/` structure (PEP 420)
   - Prevents accidental imports from source tree
   - Ensures tests run against installed package

2. **Modern pyproject.toml**: No setup.py needed
   - PEP 518 (build system requirements)
   - PEP 621 (project metadata)

3. **Optional dependencies**: Modular installation
   - Core dependencies minimal
   - Optional groups for different use cases

4. **Type hints**: Includes py.typed marker
   - Enables type checking in downstream code
   - Compatible with mypy, pyright, etc.

5. **Pre-trained models included**: Models shipped with package
   - Available after installation
   - No separate download required

## Benefits of This Setup

- ✅ Standard pip installation: `pip install elmneuron`
- ✅ Editable development mode: `pip install -e .`
- ✅ Optional dependencies: Choose what you need
- ✅ Type hints supported: Better IDE integration
- ✅ Modern standards: PEP 517/518/621 compliant
- ✅ Pre-commit hooks: Consistent code quality
- ✅ Wheel distribution: Fast installation
- ✅ Proper namespace: Import as `elmneuron`

## Troubleshooting

**Import errors after installation:**
```bash
# Verify installation
pip list | grep elmneuron

# Reinstall in editable mode
pip install -e .
```

**Build failures:**
```bash
# Update build tools
pip install --upgrade build setuptools wheel

# Clean build artifacts
rm -rf build/ dist/ *.egg-info
```

**Pre-commit hook failures:**
```bash
# Run formatters manually
black src/ tests/
isort src/ tests/

# Skip hooks if needed (not recommended)
git commit --no-verify
```

## Next Steps

1. ✅ Package is ready for local installation and testing
2. ⏳ Consider publishing to TestPyPI for validation
3. ⏳ Add CI/CD pipeline (GitHub Actions) for automated testing
4. ⏳ Set up automated package building on releases
5. ⏳ Add badges to README (build status, coverage, PyPI version)
6. ⏳ Publish to PyPI when ready for public release

## References

- [Python Packaging User Guide](https://packaging.python.org/)
- [PEP 517 - Build System Interface](https://peps.python.org/pep-0517/)
- [PEP 518 - Build System Requirements](https://peps.python.org/pep-0518/)
- [PEP 621 - Project Metadata](https://peps.python.org/pep-0621/)
- [Setuptools Documentation](https://setuptools.pypa.io/)

### CODE REVIEW

**Comment by SomeUserName1** [comment]
Changes to be confirmed by reproducing the results.
